dask_cuda/explicit_comms/comms.py

```diff
@@ -0,0 +1,184 @@
+        ep = await ucp.create_endpoint(parse_host_port(address)[0], ucx_port)
+        await ep.send(np.array([session_state["rank"]], dtype=np.uint64))
+        session_state["eps"][rank] = ep
+
+    while len(session_state["eps"]) < session_state["nworkers"] - 1:
+        await asyncio.sleep(0.1)
+
+
+async def _stop_ucp_listeners(session_state):
+    assert len(session_state["eps"]) == session_state["nworkers"] - 1
+    assert session_state["loop"] is asyncio.get_event_loop()
+    del session_state["lf"]
+
+
+async def _get_worker_state(session_state):
+    s = session_state.copy()
+    s["eps"] = [(rank, hex(ep.uid)) for rank, ep in s["eps"].items()]
+    return s["rank"]
+
+
+class CommsContext:
+    def __init__(self, client=None):
+        self.client = client if client is not None else default_client()
+        self.sessionId = uuid.uuid4().bytes
+
+        # Get address of all workers (not Nanny addresses)
+        self.worker_address = list(self.client.run(lambda: 42).keys())
+
+        ucx_ports = []
+        for rank, address in enumerate(self.worker_address):
+            ucx_ports.append(
+                self.submit(
+                    address,
+                    _create_ucp_listeners,
+                    len(self.worker_address),
+                    rank,
+                    wait=True,
+                )
+            )
+
+        # Create list of rank, address and ucx port
+        workers = [
+            (rank, address, port)
+            for rank, (address, port) in enumerate(zip(self.worker_address, ucx_ports))
+        ]
+
+        # Each worker creates a UCX endpoint to all workers with greater rank
+        futures = []
+        for i in range(len(workers)):
+            futures.append(
+                self.submit(workers[i][1], _create_endpoints, workers[i + 1 :])
+            )
+        self.client.gather(futures)
+
+        # At this point all workers should have a rank and endpoints to
+        # all other workers thus we can now stop the listening.
+        self.run(_stop_ucp_listeners)
+
+    def submit(self, worker, coroutine, *args, wait=False):
+        ret = self.client.submit(
+            _run_coroutine_on_worker,
+            self.sessionId,
+            coroutine,
+            args,
+            workers=[worker],
+            pure=False,
+        )
+        return ret.result() if wait else ret
+
+    def run(self, coroutine, *args, workers=None):
+        if workers is None:
+            workers = self.worker_address
+        ret = []
+        for worker in workers:
+            ret.append(
+                self.client.submit(
+                    _run_coroutine_on_worker,
+                    self.sessionId,
+                    coroutine,
+                    args,
+                    workers=[worker],
+                    pure=False,
+                )
+            )
+        return self.client.gather(ret)
+
+    def dataframe_operation(self, coroutine, df1, df2):
```

Comment:
Perhaps this should be more flexible, and accept a single dataframe/series as well?  I know this is only supporting merge operations for now, but i can imagine doing something similar for `set_index`/sorting.

Reply:
Agree, the plan is to support any number of inputs and more than just dataframes.